Add autostart extension

bn-darindouglass-zz opened the pull request:
- Unlike other extensions, this one provides a new `defstate` macro that
  wraps `mount.lite/defstate`s in auto-start functionality.
- Synchronize startups across threads just in case derefs happen in
  multiple threads.
- `set-autostart-fn!` allows specifying of different start
  functionalities (i.e. standard `mount.lite/start` vs extension starts)

barracudarin committed:
Add autostart extension

- Unlike other extensions, this one provides a new `defstate` macro that
  wraps `mount.lite/defstate`s in auto-start functionality.
- Synchronize startups across threads just in case derefs happen in
  multiple threads.
- `set-autostart-fn!` allows specifying of different start
  functionalities (i.e. standard `mount.lite/start` vs extension starts)

barracudarin committed:
add tests

aroemers commented:
I see this is still a draft. I think it already looks good (some tiny remarks). Do you want to open it for review?

aroemers commented on src/mount/extensions/autostart.clj:
```diff
@@ -0,0 +1,49 @@
+(ns mount.extensions.autostart
```
Could you add a small description to the namespace, for documentation purposes?

bn-darindouglass-zz replied:
Added.

aroemers commented on src/mount/extensions/autostart.clj:
```diff
@@ -0,0 +1,49 @@
+        @state))))
+
+(defmacro defstate
+  "Defines a state that should be auto-started on first deref.
```
```suggestion
  "Defines a state that will be auto-started on first deref, including its dependencies.
```

aroemers commented on test/mount/extensions/autostart_test.clj:
```diff
@@ -0,0 +1,34 @@
+    (mount/stop)
+    (is (fully? :stopped)))
+  (testing "with custom autostart-fn"
+    (sut/set-autostart-fn! deps/start)
+    (= 462 @third)
+    (is (fully? :started))
+    (mount/stop)
+    (is (fully? :stopped))))
```
I think the autostart-fn should be reset, in case the tests are run a second time within the same REPL session for example.
```suggestion
    (try
      (sut/set-autostart-fn! deps/start)
      (= 462 @third)
      (is (fully? :started))
      (mount/stop)
      (is (fully? :stopped))
      (finally
        (sut/set-autostart-fn! mount/start)))))
```

aroemers commented on test/mount/extensions/autostart_test.clj:
```diff
@@ -0,0 +1,34 @@
+  [state]
+  (= {#'first state #'second state #'third state} (mount/status)))
+
+(deftest autostart-test
```
If I read correctly, I don't think these two `testing` blocks actually test the difference between the default autostart-fn and a custom one? In both cases all three states are started. 

Maybe the `third` defstate should only depend on the `first` defstate, both explicitly as implicitly in the `:start` expression. This way, the default behaviour will start all states, whereas the custom behaviour will only start `first` and `third`.

bn-darindouglass-zz replied:
Good catch. Making the `explicit-deps` test case different from the standard `mount/start` case is a better test. Updated

aroemers reviewed (changes requested):
Looks good, only some minor comments.

bn-darindouglass-zz committed:
Update src/mount/extensions/autostart.clj

Co-authored-by: Arnout Roemers <1654946+aroemers@users.noreply.github.com>

bn-darindouglass-zz committed:
Update test/mount/extensions/autostart_test.clj

Co-authored-by: Arnout Roemers <1654946+aroemers@users.noreply.github.com>

barracudarin committed:
add namespace docstring

barracudarin committed:
make two test-cases more unique

- starting with explicit deps now starts a different state set than
  normal mount/start.

aroemers commented:
Had to fix the tests, but it's all working now. Released at 2.2.0. Thanks @DarinDouglass.
